GG/src/TextControl.cpp

```diff
@@ -204,6 +204,40 @@ void TextControl::SetText(const std::string& str)
     m_cached_minusable_size_width = X0;
 }
 
+void TextControl::SetText(const std::string& str,
+                          const std::vector<boost::shared_ptr<Font::TextElement> >& text_elements)
+{
+    if (!utf8::is_valid(str.begin(), str.end()))
+        return;
+    m_text = str;
+
+    m_code_points = CPSize(utf8::distance(str.begin(), str.end()));
+
+    m_text_elements = text_elements;
+    for (std::vector<boost::shared_ptr<Font::TextElement> >::iterator it = m_text_elements.begin();
+         it != m_text_elements.end(); ++it)
+    {
+        (*it)->Bind(m_text);
+    }
+
+    if (!m_font)
+        return;
+
+    m_line_data = m_font->DetermineLines(m_text, m_format, ClientSize().x, m_text_elements);
```

Comment:
presumably a later change will avoid redoing this for multiple uses of the same text and text elements?